rpc_util.go

```diff
@@ -307,35 +312,49 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
 			outPayload.Data = b
 			outPayload.Length = len(b)
 		}
-		if cp != nil {
-			if err := cp.Do(cbuf, b); err != nil {
-				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
+		var cbuf *bytes.Buffer
+		if cpWriter != nil || cp != nil {
+			// pre-alloc place
+			cbuf = bytes.NewBuffer(make([]byte, payloadLen+sizeLen))
+			// Has compressor, check Compressor is set by UseCompressor first.
+			if cpWriter != nil {
+				z, _ := cpWriter.Compress(cbuf)
+				if _, err := z.Write(b); err != nil {
+					return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
+				}
+				z.Close()
+			} else {
+				// If Compressor is not set by UseCompressor, use default Compressor
+				if err := cp.Do(cbuf, b); err != nil {
+					return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
+				}
 			}
 			b = cbuf.Bytes()
+			length = uint(len(b) - (payloadLen + sizeLen))
+		} else {
+			length = uint(len(b))
 		}
-		length = uint(len(b))
 	}
-	if length > math.MaxUint32 {
+
+	if length > math.MaxUint64 {
 		return nil, Errorf(codes.ResourceExhausted, "grpc: message too large (%d bytes)", length)
 	}
-
-	const (
-		payloadLen = 1
-		sizeLen    = 4
-	)
-
-	var buf = make([]byte, payloadLen+sizeLen+len(b))
-
-	// Write payload format
-	if cp == nil {
-		buf[0] = byte(compressionNone)
-	} else {
+	var buf []byte
+	if cpWriter != nil || cp != nil {
+		// re-use b
+		buf = b
```

Comment:
I was thinking that if there is no compressor, copy(buf[5:], b) will be b=buf, then, copy.
buf = append(make(byte[],5), buf) lead the size of buf uncontrolable.